src/snprintf.c

```diff
@@ -0,0 +1,63 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcutils/snprintf.h"
+
+#ifdef _WIN32
+#include <errno.h>
+#include <limits.h>
+#include <string.h>
+#endif
+#include <stdarg.h>
+#include <stdio.h>
+
+int
+rcutils_snprintf(char * buffer, size_t buffer_size, const char * format, ...)
+{
+  va_list args;
+  va_start(args, format);
+  int ret = rcutils_vsnprintf(buffer, buffer_size, format, args);
+  va_end(args);
+  return ret;
+}
+
+int
+rcutils_vsnprintf(char * buffer, size_t buffer_size, const char * format, va_list args)
+{
+  if (
+    !(NULL == buffer && 0 == buffer_size) &&
```

Comment:
I imagine the purpose of this is to allow the use case of passing (null,0,format,..) in order to get the number of characters that would-be formatted (as is done [here](https://github.com/ros2/rcutils/pull/85/files#diff-d20fbd64fcff93291e6ef221061e976aR51)). However, I still think we should ensure that `format` is not null in that case.

On windows, https://msdn.microsoft.com/en-us/library/d3xd30zz.aspx suggests to me this use case of calculating would-be characters is not supported by `_vsnprintf_s` (it will set `errno` if any are null, and we will end up returning -1 below). It would be good for portability to instead use `_vscprintf` in that case, so that `rcutils_vsnprintf` can always be used with a null buffer, at which point we can replace checks [such as this](https://github.com/ros2/rcutils/pull/85/files#diff-d20fbd64fcff93291e6ef221061e976aR48) with a single call to `rcutils_vsnprintf`. Could be left for another day though since we already have this disparity.

Reply:
That's a good point (or a couple). I'll fix it up some and get you to look again.